Packages/ClientRuntime/Sources/Waiters/WaiterRetryer.swift

```diff
@@ -0,0 +1,88 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+
+/// A type which tries the waited operation after a delay, then returns or throws the result.
+/// Called by the `Waiter` to make the initial request, then called again for an additional request when retry
+/// is needed.
+/// A custom closure may be injected at initialization to allow `WaiterRetryer` to be used as a mock.
+class WaiterRetryer<Input, Output> {
+    typealias WaitThenRequest =
+        (WaiterScheduler, Input, WaiterConfiguration<Input, Output>, (Input) async throws -> Output)
+            async throws -> WaiterOutcome<Output>?
+
+    private let closure: WaitThenRequest
```

Comment:
Nit: prefer a descriptive name over `closure`. 
For example, we wouldn't have name a property `string` or `bool` etc

Reply:
I'm not too concerned because this is not a public interface, but it is renamed to `waitThenRequest` like the function it wraps.